epics/pv.py

```diff
@@ -356,10 +356,22 @@ def _set_charval(self, val, call_ca=True):
         # char waveform as string
         if ntype == dbr.CHAR and self.count < ca.AUTOMONITOR_MAXLENGTH:
             if isinstance(val, ca.numpy.ndarray):
+                # a numpy array
                 val = val.tolist()
-            elif self.count==1: # handles single character in waveform
-                val = [val]
-            val = list(val)
+
+                if not isinstance(val, list):
+                    # a scalar value from numpy, tolist() turns it into a
+                    # native python integer
```

Comment:
As the comment mentions, we can sometimes get a single scalar value back from `tolist()`:
```python
In [1]: import numpy as np

In [2]: scalar_value = np.float64(1)

In [3]: scalar_value.tolist()
Out[3]: 1.0

In [4]: type(scalar_value.tolist())
Out[4]: float
```

This breaks the ability to search for a null terminator, hence the check for the list type here.